dali_tf_plugin/setup.py.in

```diff
@@ -14,16 +14,64 @@
 
 from setuptools import setup, find_namespace_packages
 from dali_tf_plugin_install_tool import InstallerHelper
-from setuptools.command.build_ext import build_ext
 from setuptools.dist import Distribution
 from setuptools.extension import Extension
 import os
+import sys
+
+from setuptools.command.build_ext import build_ext
+from setuptools.command.egg_info import egg_info
+
+DALI_PKG_REQ = 'nvidia-dali@DALI_FLAVOR_MINUS@-cuda@CUDA_VERSION_SHORT_DIGIT_ONLY@==@DALI_VERSION@'
+NVIDIA_INDEX_URL = "https://developer.download.nvidia.com/compute/redist/@DALI_FLAVOR@"
+
+def is_dali_installed():
+    try:
+        import nvidia.dali
+        return True
+    except:
+        return False
+
+def pip_install(pkg_req, user=False):
+    pip_args = [
+        "install",
+        "--index-url", NVIDIA_INDEX_URL
+    ]
+    if user:
+        pip_args.append("--user")
+    pip_args.append(pkg_req)
+
+    from pip._internal import main as pip
+    return pip(pip_args) == 0
 
 class CustomBuildExt(build_ext, object):
     def run(self):
+        if not is_dali_installed():
+            raise RuntimeError("NVIDIA DALI should be installed at this point.")
         helper = InstallerHelper(plugin_dest_dir = os.path.join(self.build_lib, 'nvidia', 'dali_tf_plugin'))
         helper.install()
 
+class EggInfoCommand(egg_info):
+    """A class for setuptools egg_info command.
+    DALI TF requires DALI to be installed when the package is built (not installed).
+    We are overriding egg_info command to install DALI package before build_ext command,
+    if it was not installed already.
+    """
+
+    def run(self):
+        """Installing DALI if not already installed during EggInfo command (first egg_info run)"""
+        handle_deps = all([keyword in sys.argv for keyword in ["egg_info", "--egg-base"]]) 
+        if handle_deps and not is_dali_installed():
+            ret = pip_install(DALI_PKG_REQ, user=False)
+            if not ret:
+                ret = pip_install(DALI_PKG_REQ, user=True)
+            if not ret:
+                raise RuntimeError("Could not install {}".format(DALI_PKG_REQ))
+            # For nvidia.dali to be importable, restart the program from scratch
+            os.execl(sys.executable, sys.executable, *sys.argv)
```

Comment:
Maybe it is enough to reaload `setuptools` or other modules (see https://www.geeksforgeeks.org/reloading-modules-python)?

Reply:
reload is to reload a module that was already imported. In this case, when I try to import `nvidia.dali` I get an error saying that there's no such package. When you rerun the program the package is available. I am assuming that when we start python the available packages are loaded into memory. Not sure if we can trigger reloading that, so far I haven't found any info about it.